src/dag.rs

```diff
@@ -263,47 +263,186 @@ impl<J: jet::Jet> DagLike for Rc<RedeemNode<J>> {
+                            current.left_idx = Some(lidx);
+                            current.right_idx = Some(ridx);
+                            self.stack.push(current);
+                        }
+                        // Two children, one yielded already
+                        (Child::New(child), Child::Repeat { idx }) => {
+                            current.right_idx = Some(idx);
+                            self.stack.push(current);
+                            self.stack
+                                .push(IterStackItem::unprocessed(child, Previous::ParentLeft));
+                        }
+                        (Child::Repeat { idx }, Child::New(child)) => {
+                            current.left_idx = Some(idx);
+                            self.stack.push(current);
+                            self.stack
+                                .push(IterStackItem::unprocessed(child, Previous::ParentRight));
+                        }
+                        // Two children, neither yielded already
+                        (Child::New(lchild), Child::New(rchild)) => {
+                            self.stack.push(current);
+                            self.stack
+                                .push(IterStackItem::unprocessed(rchild, Previous::ParentRight));
+                            self.stack
+                                .push(IterStackItem::unprocessed(lchild, Previous::SiblingLeft));
+                        }
                     }
-                }
-                // else
-                self.maybe_current = None;
-            } else if let Some(top) = self.stack.last() {
-                if let Some(right) = top.right_child() {
-                    if !self.visited.contains(&PointerId::from(&right)) {
-                        self.maybe_current = Some(right);
-                        continue;
+                } else {
+                    // The next time we encounter an item, we have the indices for its children
+                    // (which we ignore for now, but will use in a future commit) and can yield
+                    // it. But first we need to update the indices of its parent.
+                    let stack_len = self.stack.len();
+                    match current.previous {
+                        Previous::Root => {
+                            assert_eq!(stack_len, 0);
+                        }
+                        Previous::ParentLeft => {
+                            assert!(self.stack[stack_len - 1].processed);
+                            self.stack[stack_len - 1].left_idx = Some(self.index);
+                        }
+                        Previous::ParentRight => {
+                            assert!(self.stack[stack_len - 1].processed);
+                            self.stack[stack_len - 1].right_idx = Some(self.index);
+                        }
+                        Previous::SiblingLeft => {
+                            assert!(self.stack[stack_len - 2].processed);
```

Comment:
I know we have a lot of `assert!`s, but what is our current policy? Should `assert!` become `debug_assert!`?